Core/src/Vertexing/KalmanVertexUpdater.cpp

```diff
@@ -190,30 +221,212 @@ void update(Vector4& vtxPos, SquareMatrix4& vtxCov,
+  const auto correctedPhiTheta =
+      Acts::detail::normalizePhiTheta(newTrkMomentum(0), newTrkMomentum(1));
+  newTrkParams(BoundIndices::eBoundPhi) = correctedPhiTheta.first;     // phi
+  newTrkParams(BoundIndices::eBoundTheta) = correctedPhiTheta.second;  // theta
+  newTrkParams(BoundIndices::eBoundQOverP) = newTrkMomentum(2);        // qOverP
+
+  // E_k^n
+  const ActsMatrix<nDimVertex, 3> crossCovVP =
+      -vtxCov * posJac.transpose() * trkParamWeight * momJac * cache.wMat;
+
+  // Difference in positions. cache.newVertexPos corresponds to \tilde{x_k^{n*}} in Ref. (1).
+  VertexVector posDiff =
+      vtxPos - cache.newVertexPos.template head<nDimVertex>();
+
+  // r_k^n
+  ParameterVector paramDiff =
+      trkParams - (constTerm + posJac * vtxPos + momJac * newTrkMomentum);
+
+  // New chi2 to be set later
+  double chi2 =
+      posDiff.dot(
+          cache.newVertexWeight.template block<nDimVertex, nDimVertex>(0, 0) *
+          posDiff) +
+      paramDiff.dot(trkParamWeight * paramDiff);
+
+  Acts::BoundMatrix newTrackCov = calculateTrackCovariance<nDimVertex>(
+      cache.wMat, crossCovVP, vtxCov, newTrkParams);
+
+  // Create new refitted parameters
+  std::shared_ptr<PerigeeSurface> perigeeSurface =
+      Surface::makeShared<PerigeeSurface>(vtxPos.template head<3>());
+
+  BoundTrackParameters refittedPerigee =
+      BoundTrackParameters(perigeeSurface, newTrkParams, std::move(newTrackCov),
+                           track.fittedParams.particleHypothesis());
+
+  // Set new properties
+  track.fittedParams = refittedPerigee;
+  track.chi2Track = chi2;
+  track.ndf = 2 * track.trackWeight;
+
+  return;
 }
 
-void KalmanVertexUpdater::detail::calculateUpdate4(
-    const Vector4& vtxPos, const SquareMatrix4& vtxCov,
-    const Acts::LinearizedTrack& linTrack, const double trackWeight,
-    const int sign, Cache<4>& cache) {
-  calculateUpdateImpl<4>(vtxPos, vtxCov, linTrack, trackWeight, sign, cache);
+}  // namespace
+
+// The two functions don't contain any of the actual update code, they
+// only dispatch into templated functions, effectively doing a
+// runtime-to-compile time conversion.
+
+void KalmanVertexUpdater::updateVertexWithTrack(Vertex& vtx, TrackAtVertex& trk,
+                                                unsigned int nDimVertex) {
+  if (nDimVertex == 3) {
+    updateVertexWithTrackImpl<3>(vtx, trk, 1);
+  } else if (nDimVertex == 4) {
+    updateVertexWithTrackImpl<4>(vtx, trk, 1);
+  } else {
+    throw std::invalid_argument(
+        "The vertex dimension must either be 3 (when fitting the spatial "
+        "coordinates) or 4 (when fitting the spatial coordinates + time).");
+  }
 }
 
-void KalmanVertexUpdater::detail::updateVertexWithTrack(
-    Vector4& vtxPos, SquareMatrix4& vtxCov,
-    std::pair<double, double>& fitQuality, TrackAtVertexRef trk, int sign,
-    unsigned int nDimVertex) {
+void Acts::KalmanVertexUpdater::updateTrackWithVertex(TrackAtVertexRef track,
```

Comment:
```suggestion
void Acts::KalmanVertexUpdater::updateTrackWithVertex(TrackAtVertex& track,
```
That might improve readability. I think there are a couple other places and we could also remove the typedef

Reply:
Good point, I might not need this dance anymore either. I'll check!

Reply:
Removed now.